src/transform-connection.js

```diff
@@ -33,15 +34,41 @@ function contextsFromNearestNode(context) {
   }
 }
 
-function addNextFieldTo(currentSelection, contextChain, cursor, path) {
+function initializeDocumentAndVars(currentContext, contextChain) {
+  const lastInChain = contextChain[contextChain.length - 1];
+  const first = lastInChain.selection.args.first;
+  const variableDefinitions = Object
+    .keys(lastInChain.selection.args)
+    .filter((key) => {
+      return VariableDefinition.prototype.isPrototypeOf(lastInChain.selection.args[key]);
+    })
+    .map((key) => {
+      return lastInChain.selection.args[key];
+    });
+
+  let firstVar = variableDefinitions.find((definition) => {
+    return definition.name === 'first';
+  });
+
+  if (!firstVar) {
+    firstVar = variable('first', 'Int', first);
+    variableDefinitions.push(firstVar);
+  }
+
+  const document = new Document(currentContext.selection.selectionSet.typeBundle);
+
+  return [document, variableDefinitions, firstVar];
+}
+
+function addNextFieldTo(currentSelection, contextChain, path, cursor) {
   // There are always at least two. When we start, it's the root context, and the first set
   const nextContext = contextChain.shift();
 
   path.push(nextContext.selection.responseKey);
 
   if (contextChain.length) {
     currentSelection.add(nextContext.selection.name, {alias: nextContext.selection.alias, args: nextContext.selection.args}, (newSelection) => {
-      addNextFieldTo(newSelection, contextChain, cursor, path);
+      addNextFieldTo(newSelection, contextChain, path, cursor);
```

Comment:
was this just a bug in positions before? Might be worthwhile to make these named property based args if so.

Reply:
Not really. The first three args are dynamic. The cursor part is static. In one iteration, I grouped `cursor` and other static parts in a hash so they're more like config, but then I didn't need the last argument. If I added anything more, I'd wrap them up. As it stands, this method is only used recursively, and in this file, so doing `{newSelection, contextChain, path, cursor}` doesn't really make the signature any more clear.